app/gui/src/project-view/stores/project/index.ts

```diff
@@ -387,15 +384,31 @@ export function createProjectStore(
         reject(Err(message))
       }
 
-      wait((timeoutMs < 1000 ? 1000 : timeoutMs) + 100).then(() => {
-        if (state === 1) {
-          inProgress.value -= 1
-          state = 0 // Prevent further updates from this handler.
-          dataConnection.off(`${OutboundPayload.VISUALIZATION_UPDATE}`, dataHandler)
-          executionContext.off('visualizationEvaluationFailed', errorHandler)
-          reject(Err(`executeExpression: Execution timed out.`))
-        }
-      })
+      const waitWithExponentialBackoff = (retryAttempt: number, timeoutMs: number) => {
+        wait(timeoutMs).then(() => {
+          if (state === 1) {
+            if (retryAttempt < MAX_RETRIES_IN_QUEUE) {
+              const incRetryAttempt = retryAttempt + 1
+              console.warn(
+                'Waiting on data (expressionId=' +
+                  expressionId +
+                  ', visualizationId=' +
+                  visualizationId +
+                  '), retry attempt: ' +
+                  incRetryAttempt,
+              )
+              waitWithExponentialBackoff(incRetryAttempt, timeoutMs * 2 ** retryAttempt)
+            } else {
+              inProgress.value -= 1
+              state = 0 // Prevent further updates from this handler.
+              dataConnection.off(`${OutboundPayload.VISUALIZATION_UPDATE}`, dataHandler)
+              executionContext.off('visualizationEvaluationFailed', errorHandler)
+              reject(Err(`executeExpression: Execution timed out.`))
+            }
+          }
+        })
+      }
+      waitWithExponentialBackoff(0, timeoutMs)
```

Comment:
This seems like a lot of complexity to schedule printing some debug messages.

I have some questions:
- Do we need this message at all? If I see it in the console I wouldn't know what I'm supposed to do about it.
- Does it need to be warning-level, with a stack trace?
- Do we need it in release builds, or could it be gated with `DEV:`?
- Would it not provide equally useful information to wait once, and in the log message include the elapsed time at success/cancelation?

Reply:
True, this should be a dev warning.